File: scripts/generate_dev_secrets.py
#!/usr/bin/env python3
"""
Generate .env from fullstack template. Template uses __VAR_NAME__ placeholders
for secrets (env var or generated).

Source: scripts/env/.env.template (fullstack: infra + app + auth).

Usage:
  python scripts/generate_dev_secrets.py              # generate .env from template
  python scripts/generate_dev_secrets.py --verify      # only validate existing .env
  OPENAI_API_KEY=xxx python scripts/generate_dev_secrets.py
"""
from __future__ import annotations

import argparse
import os
import re
import secrets
import sys
from pathlib import Path


def root_dir() -> Path:
    return Path(__file__).resolve().parent.parent


SECRET_KEYS = [
    "KEYCLOAK_ADMIN_PASSWORD",
    "KEYCLOAK_DB_PASSWORD",
    "QDRANT_API_KEY",
    "SESSION_SECRET",
    "OPENAI_API_KEY",
]

PLACEHOLDER_RE = re.compile(r"^__([A-Za-z_][A-Za-z0-9_]*)__$")


def is_placeholder(value: str) -> bool:
    return bool(value and PLACEHOLDER_RE.match(value.strip()))


def parse_env_file(path: Path) -> dict[str, str]:
    """Parse KEY=value lines; skip comments and empty. Returns dict key -> value."""
    out: dict[str, str] = {}
    if not path.exists():
        return out
    for line in path.read_text().splitlines():
        line = line.strip()
        if not line or line.startswith("#"):
            continue
        m = re.match(r"^([A-Za-z_][A-Za-z0-9_]*)=(.*)$", line)
        if m:
            out[m.group(1)] = m.group(2)
    return out


def read_env_lines(path: Path) -> list[str]:
    """Read file as lines (preserve order and comments)."""
    if not path.exists():
        return []
    return path.read_text().splitlines()


def write_env_file(path: Path, data: dict[str, str], key_order: list[str] | None = None) -> None:
    """Write KEY=value file. If key_order given, output in that order; else sorted."""
    path.parent.mkdir(parents=True, exist_ok=True)
    if key_order:
        items = [(k, data[k]) for k in key_order if k in data]
        for k in sorted(data):
            if k not in {x[0] for x in items}:
                items.append((k, data[k]))
    else:
        items = sorted(data.items())
    lines = [f"{k}={v}" for k, v in items]
    path.write_text("\n".join(lines) + "\n")


def resolve_secrets(keys: list[str], existing: dict[str, str], force: bool) -> dict[str, str]:
    """For each key, return env or existing or generated secret."""
    resolved: dict[str, str] = {}
    for key in keys:
        val = os.environ.get(key) or existing.get(key)
        if val and not force:
            resolved[key] = val
        else:
            if key in ("KEYCLOAK_ADMIN_PASSWORD", "KEYCLOAK_DB_PASSWORD"):
                resolved[key] = secrets.token_urlsafe(24)
            elif key == "QDRANT_API_KEY":
                resolved[key] = secrets.token_urlsafe(24)
            elif key == "SESSION_SECRET":
                resolved[key] = os.environ.get(key) or secrets.token_hex(32)
            elif key == "OPENAI_API_KEY":
                resolved[key] = os.environ.get(key) or ""
            else:
                resolved[key] = os.environ.get(key) or ""
    return resolved


def replace_placeholders(data: dict[str, str], secrets_map: dict[str, str]) -> dict[str, str]:
    """Replace any value __KEY__ with actual value from secrets or env."""
    out: dict[str, str] = {}
    for k, v in data.items():
        v = (v or "").strip()
        if is_placeholder(v):
            var_name = PLACEHOLDER_RE.match(v).group(1)
            out[k] = secrets_map.get(var_name) or os.environ.get(var_name) or ""
        else:
            out[k] = v
    return out


def main() -> None:
    ap = argparse.ArgumentParser(description="Generate .env from fullstack template.")
    ap.add_argument("--verify", action="store_true", help="Only validate existing .env")
    ap.add_argument("--force", action="store_true", help="Regenerate secrets even if present")
    args = ap.parse_args()

    root = root_dir()
    env_file = root / ".env"
    env_dir = root / "scripts" / "env"
    env_tpl = env_dir / ".env.template"

    if args.verify:
        if not env_file.exists():
            raise SystemExit(".env: file not found")
        data = parse_env_file(env_file)
        required = SECRET_KEYS + ["REDIS_URL", "QDRANT_URL", "QDRANT_COLLECTION"]
        missing = [k for k in required if k not in data or not (data.get(k) or "").strip()]
        if missing:
            raise SystemExit(f".env: missing or empty: {', '.join(missing)}")
        print("Verify OK: .env has required keys.")
        return

    if not env_tpl.exists():
        raise SystemExit("Need scripts/env/.env.template (fullstack) to generate .env")

    existing = parse_env_file(env_file)
    secrets_map = resolve_secrets(SECRET_KEYS, existing, args.force)
    data = parse_env_file(env_tpl)
    data = replace_placeholders(data, secrets_map)
    for k in SECRET_KEYS:
        if k not in data or not (data.get(k) or "").strip():
            data[k] = secrets_map.get(k, "")

    key_order = [line.strip().split("=")[0] for line in read_env_lines(env_tpl) if re.match(r"^[A-Za-z_][A-Za-z0-9_]*=", line.strip())]
    write_env_file(env_file, data, key_order=key_order if key_order else None)
    print("Wrote .env (fullstack: infra + app + auth).")

    data = parse_env_file(env_file)
    missing = [k for k in SECRET_KEYS + ["REDIS_URL", "QDRANT_URL", "QDRANT_COLLECTION"] if k not in data or not (data.get(k) or "").strip()]
    if missing:
        raise SystemExit(f".env: missing or empty after write: {', '.join(missing)}")

    print("Done. Use: docker compose --env-file .env --profile infra up -d")


if __name__ == "__main__":
    main()
    sys.exit(0)

File: scripts/generate_dev_secrets.mjs
#!/usr/bin/env node
/**
 * Generate .env from template. Template uses __VAR_NAME__ placeholders
 * replaced by env var or generated secret.
 * Source of truth: scripts/env/.env.template
 * Outputs: .env (all enabled). Plain "npm test" overrides AUTH_ENABLED=false REDIS_URL=.
 * Usage: node scripts/generate_dev_secrets.mjs [--write-templates] [--verify] [--force]
 */
import { readFileSync, writeFileSync, existsSync, mkdirSync, unlinkSync } from 'fs';
import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { randomBytes } from 'crypto';

const __dirname = dirname(fileURLToPath(import.meta.url));
const root = join(__dirname, '..');

const PROD_PORT_DEFAULTS = { PORT: 3000, METRICS_PORT: 9090, QDRANT_HTTP: 6333, REDIS: 6379, KEYCLOAK: 8080 };
const DEV_OFFSET = 300;
const SECRET_KEYS = ['KEYCLOAK_ADMIN_PASSWORD', 'KEYCLOAK_DB_PASSWORD', 'QDRANT_API_KEY', 'SESSION_SECRET', 'OPENAI_API_KEY'];
const PLACEHOLDER_RE = /^__([A-Za-z_][A-Za-z0-9_]*)__$/;

function parseEnvFile(path) {
  const out = {};
  if (!existsSync(path)) return out;
  for (const line of readFileSync(path, 'utf8').split('\n')) {
    const s = line.trim();
    if (!s || s.startsWith('#')) continue;
    const m = s.match(/^([A-Za-z_][A-Za-z0-9_]*)=(.*)$/);
    if (m) out[m[1]] = m[2];
  }
  return out;
}

function readEnvLines(path) {
  if (!existsSync(path)) return [];
  return readFileSync(path, 'utf8').split('\n');
}

function resolveSecrets(keys, existing, force) {
  const resolved = {};
  for (const key of keys) {
    let val = process.env[key] || existing[key];
    if (val && !force) {
      resolved[key] = val;
      continue;
    }
    if (['KEYCLOAK_ADMIN_PASSWORD', 'KEYCLOAK_DB_PASSWORD', 'QDRANT_API_KEY'].includes(key)) {
      resolved[key] = randomBytes(24).toString('base64url');
    } else if (key === 'SESSION_SECRET') {
      resolved[key] = process.env[key] || randomBytes(32).toString('hex');
    } else if (key === 'OPENAI_API_KEY') {
      resolved[key] = process.env[key] || '';
    } else {
      resolved[key] = process.env[key] || '';
    }
  }
  return resolved;
}

function applyDevPorts(data) {
  const port = PROD_PORT_DEFAULTS.PORT + DEV_OFFSET;
  const metrics = PROD_PORT_DEFAULTS.METRICS_PORT + DEV_OFFSET;
  const qdrant = PROD_PORT_DEFAULTS.QDRANT_HTTP + DEV_OFFSET;
  const redisPort = PROD_PORT_DEFAULTS.REDIS + DEV_OFFSET;
  const keycloak = PROD_PORT_DEFAULTS.KEYCLOAK + DEV_OFFSET;
  data.PORT = String(port);
  data.METRICS_PORT = String(metrics);
  data.AUTH_CALLBACK_BASE_URL = `http://localhost:${port}`;
  data.QDRANT_URL = `http://localhost:${qdrant}`;
  data.REDIS_URL = `redis://127.0.0.1:${redisPort}`;
  data.KEYCLOAK_URL = `http://localhost:${keycloak}`;
  data.AUTH_TRUSTED_ISSUERS = `http://localhost:${keycloak}/realms/kairos-dev`;
}

function isPlaceholder(v) {
  return v && PLACEHOLDER_RE.test(String(v).trim());
}

function replacePlaceholders(data, secretsResolved) {
  const out = {};
  for (const [k, v] of Object.entries(data)) {
    const s = (v || '').trim();
    if (isPlaceholder(s)) {
      const name = s.match(PLACEHOLDER_RE)[1];
      out[k] = secretsResolved[name] ?? process.env[name] ?? '';
    } else {
      out[k] = v;
    }
  }
  return out;
}

function writeEnvFile(path, data, keyOrder = null) {
  mkdirSync(dirname(path), { recursive: true });
  const keys = keyOrder ? [...keyOrder.filter((k) => k in data), ...Object.keys(data).filter((k) => !keyOrder.includes(k))] : Object.keys(data).sort();
  const lines = keys.map((k) => `${k}=${data[k]}`);
  writeFileSync(path, lines.join('\n') + '\n');
}

function writeTemplateFromBkp(bkpPath, outPath, secretKeys) {
  const lines = readEnvLines(bkpPath);
  const outLines = lines.map((line) => {
    const stripped = line.trim();
    const m = stripped.match(/^([A-Za-z_][A-Za-z0-9_]*)=(.*)$/);
    if (m && secretKeys.has(m[1])) return `${m[1]}=__${m[1]}__`;
    return line;
  });
  mkdirSync(dirname(outPath), { recursive: true });
  writeFileSync(outPath, outLines.join('\n') + '\n');
}

function main() {
  const args = process.argv.slice(2);
  const writeTemplates = args.includes('--write-templates');
  const verify = args.includes('--verify');
  const force = args.includes('--force');

  const bkp = join(root, 'bkp');
  const envFile = join(root, '.env');
  const envDir = join(root, 'scripts', 'env');
  const envTpl = join(envDir, '.env.template');

  if (writeTemplates) {
    const bkpEnv = join(bkp, '.env');
    const bkpDev = join(bkp, '.env.dev');
    if (existsSync(bkpEnv) && existsSync(bkpDev)) {
      const merged = { ...parseEnvFile(bkpEnv), ...parseEnvFile(bkpDev) };
      const mergedPath = join(envDir, '.env.merged.bkp');
      writeEnvFile(mergedPath, merged);
      writeTemplateFromBkp(mergedPath, envTpl, new Set(SECRET_KEYS));
      try { unlinkSync(mergedPath); } catch (_) {}
    } else if (existsSync(bkpEnv)) {
      writeTemplateFromBkp(bkpEnv, envTpl, new Set(SECRET_KEYS));
    } else {
      process.stderr.write('bkp/.env (or bkp/.env + bkp/.env.dev) must exist to write template.\n');
      process.exit(1);
    }
    console.log('Wrote scripts/env/.env.template (secrets replaced by __VAR__).');
    return;
  }

  if (verify) {
    if (!existsSync(envFile)) {
      process.stderr.write('.env: file not found\n');
      process.exit(1);
    }
    const data = parseEnvFile(envFile);
    const required = ['REDIS_URL', 'QDRANT_URL', 'QDRANT_COLLECTION'].concat(
      SECRET_KEYS.filter((k) => ['KEYCLOAK_ADMIN_PASSWORD', 'KEYCLOAK_DB_PASSWORD'].includes(k))
    );
    const missing = required.filter((k) => !(data[k] || '').trim());
    if (missing.length) {
      process.stderr.write(`.env: missing or empty: ${missing.join(', ')}\n`);
      process.exit(1);
    }
    console.log('Verify OK: .env has required keys.');
    return;
  }

  const src = existsSync(envTpl) ? envTpl : join(bkp, '.env');
  if (!existsSync(src)) {
    process.stderr.write('Need scripts/env/.env.template or bkp/.env to generate .env\n');
    process.exit(1);
  }

  const existing = parseEnvFile(envFile);
  const secretsResolved = resolveSecrets(SECRET_KEYS, existing, force);
  let data = parseEnvFile(src);
  data = replacePlaceholders(data, secretsResolved);
  applyDevPorts(data);
  for (const k of SECRET_KEYS) {
    if (!(data[k] || '').trim()) data[k] = secretsResolved[k] ?? '';
  }

  const keyOrder = [];
  for (const line of readEnvLines(src)) {
    const s = line.trim();
    if (s.startsWith('#')) continue;
    const m = s.match(/^([A-Za-z_][A-Za-z0-9_]*)=/);
    if (m && !keyOrder.includes(m[1])) keyOrder.push(m[1]);
  }

  data.AUTH_ENABLED = 'true';
  writeEnvFile(envFile, data, keyOrder.length ? keyOrder : null);
  console.log('Wrote .env (all enabled). Use for start/dev/infra and test:integration. Plain npm test uses AUTH_ENABLED=false REDIS_URL= overrides.');

  const dataAfter = parseEnvFile(envFile);
  const missing = ['REDIS_URL', 'QDRANT_URL', 'QDRANT_COLLECTION'].filter((k) => !(dataAfter[k] || '').trim());
  if (missing.length) {
    process.stderr.write(`.env: missing or empty after write: ${missing.join(', ')}\n`);
    process.exit(1);
  }
  console.log('Done. npm start / npm run dev / npm run infra:up use .env. npm test = memory + no auth (overrides). npm run test:integration = full .env.');
}

main();
